.travis.yml

```diff
@@ -1,6 +1,12 @@
 language: python
 
-python: "2.7"
+python:
+  - "2.7"
+  - "3.6"
+
+matrix:
+  allow_failures:
+    - python: "3.6"
```

Comment:
Rather than allowing failures I would prefer to have the tests passing for Python 3.

Reply:
I have seen this mistake made too many times.  Fixing all Python 3 issues in a single PR is a mistake because it takes too long to test the whole in Python 2 (See the Facebook presos at PyCon).  For example, this PR already touches 16 files and is so complex to review that it has not progressed in the three months since it would open.

I would highly recommend that we take an _incremental_ approach to Python 3 compatibility using this sequence of actions:
1. Get the whole codebase to work on Python 2 with no Python 3 syntax errors or undefined names (we need to prove that Python 2 test pass even if Python 3 tests still fill.)
2. Get the whole codebase to work on both Python 2 and Python 3.  (All tests pass.)

We are still on step 1 and will not start on step 2 until _after_ step 1 is complete.

Reply:
> For example, this PR already touches 16 files and is so complex to review ...

That is simply not the case. As commented several of the changes are unnecessary (print future) or unrelated to Python 3 (scoped file handle). The remaining changes are trivial parenthesis for print statements as well as one conditional import. The patch is extremely easy to review.

I am sorry for the delay. The delay was mostly due to one person starting the review but then not following up on it.

Please continue to address the remaining Python 3 problems step by step.